### ISSUE TITLE
heads up: building texmath when ghc is configured to use Clang rather than GCC can cause HIGH memory usage

### ISSUE THREAD

**Issue by cartazio** [jgm/texmath#70]
somehow on certain invocations of the build process for the texmath package,  if GHC is configured to use clang rather than GCC, clang will consume >= 1gb of memory on assembling/compiling certain modules, whereas GCC seems to max out at needing ~ 1/2gb. 

I realize this is probably an upstream problem, but making you aware of it none the less, because it will likely impact some of your users

**Comment by mpickering** [comment]
@cartazio I've wanted to move these tables out of source, do you just mean having the table in a C file and using FFI?

**Comment by cartazio** [comment]
yeah, you could just FFI them as a a Pointer, then use Storables on the haskell side, happy to help

**Comment by cartazio** [comment]
remind me and I'll try to dig up some example haskell codes that do this already

**Comment by jgm** [comment]
That would be very helpful.

+++ Carter Tazio Schonwald [Jan 13 15 21:31 ]:

**Comment by winterkoninkje** [comment]
I'm also running into this problem when using gcc on OSX: the mmap when compiling Text.TeXMath.Unicode.ToTeX fails. Would disabling profiling help with that too?

**Comment by winterkoninkje** [comment]
Looks like yes: with profiling disabled it finishes quickly with gcc too

**Comment by jgm** [comment]
@cartazio, if you could supply an example that would point the way forward here, that would be great.
(My laptop is heating up now as texmath compiles...)

**Comment by cartazio** [comment]
@wrengr  gcc on mac IS clang

ok, sorry for the MIA

theres actually a "simpler" way you could hack it.  (may not be simpler per se, )
1) the compilation time blowup is presumably because of a superlinearity in the compiler complexity
2) if you split the offending list into 4-8 disjoint ranges, each in their own private internal module, and do a teeny bit of wrapper logic 

anyways, writing up the model implementation right now, and i'll email @jgm the code plus link a (truncated!) gist of the approach here

**Comment by cartazio** [comment]
oooo, it works, 
```haskell
-- {-#  LANGUAGE ForeignFunctionInterface #-}

--module Text.TeXMath.Unicode.ToASCII (getASCII) where

module ToASCII where
import qualified Data.IntMap as M
import Data.Char (ord)
import Data.Maybe (fromMaybe)
import Foreign.C
import Foreign.Ptr
import System.IO.Unsafe
import Foreign.Storable
--import Data.Int
-- | Approximates a single unicode character as an ASCII string
--  (each character is between 0x00 and 0x7F).
getASCII :: Char -> String
getASCII u = fromMaybe "" (M.lookup (ord u) table)

--table :: M.IntMap String

table :: M.IntMap String
table = M.fromList $ zip realKeyKey realValVal

foreign import ccall unsafe "&keylookup" keyKey :: Ptr CInt
foreign import ccall unsafe "&toASCIILut" valVal :: Ptr (Ptr CChar)

realKeyKey :: [Int]
realKeyKey = unsafePerformIO $ mapM (\off ->  do  daInt <- peekElemOff keyKey off ; return $ fromIntegral daInt ) [ 0 .. 8922]

realValVal :: [String]
realValVal = unsafePerformIO $ mapM (\off -> do daPtrStr <- peekElemOff valVal off ; peekCString daPtrStr ) [0 .. 8922]

```

then in ghci i do eg 
```
*ToASCII> getASCII 'µ'
"u"

```

yay!

ill turn this into a pr now

**Comment by cartazio** [comment]
note that I make this lookup table code available under a BSD/MIT license, so when other folks wanna adapt the idea to their code, they are welcome to use it as is (though this project itself is gpl now for some reason :) )

**Comment by cartazio** [comment]
`Text.TeXMath.Unicode.ToTeX` is still egregiously slow, but i leave adapting the same approach to that module to someone else :) 

so i'm gooona propose we also remove ` -auto-all -caf-all` from the default build profile options for now

**Comment by cartazio** [comment]
pr #71  does the first bits

**Comment by jgm** [comment]
Thanks!  ToTeX will be harder because of the more complex data structure.  But this is a big improvement already; my laptop no longer sounds like a vacuum cleaner while compiling texmath.

**Comment by cartazio** [comment]
Half the change Is in my PR I disabled prof auto all and friends.  :)
though I suspect the blow up still happenes on totex even then.

Once you migrate all the lookup structures to c code, it reenabling those
profiling flags will no longer blow up compile time.

**Comment by cartazio** [comment]
I'll take a look at the totex structure this evening.  Now that I've got one part working, should be super easy to adapt to the rest.

**Comment by jgm** [comment]
Great!

**Comment by cartazio** [comment]
nearly done, stay tuned :)

**Comment by cartazio** [comment]
ok, the PR for the ToTex stuff SHOULD pass this next round of travis CI

all thats left is 
1) the entity maps code
2) doing a subsequent refactor so that @mpickering's code gen approach can be migrated to this style so that its a bit easier to upgrade / mess with in the future. (i probably should have done that from the outset in retrospect, but at least you have some worked out proof of concept implemntations that demonstrate the idea :) )

**Comment by cartazio** [comment]
(if one of you wanna take over on those fronts, please do :) )

**Comment by cartazio** [comment]
another option  would have been to add those unicode files as "extra data files", but that adds to relocatability headaches so i didn't consider that route

**Comment by jgm** [comment]
@cartazio I wonder whether the changes you made are related to this report I just got?
https://github.com/jgm/pandoc/commit/693bc5f50f2607cd0d5fbd2ff7328850a6ef7236#co
mmitcomment-10723486

**Comment by cartazio** [comment]
The version bound stuff? Did he try doing a cabal update?  I'm afk, but I'll write out build plan debugging directions later

**Comment by jgm** [comment]
+++ Carter Tazio Schonwald [Apr 14 15 14:46 ]:
>The version bound stuff? Did he try doing a cabal update?  I'm afk, but I'll write out build plan debugging directions later

I should have given more background.  I recently raised the lower bound on texmath, which forced installation of a version with the changes you pushed (the C data files).  I thought this might be related to the problem he was having, though it seems that it cleared up after a cabal clean.

**Comment by jgm** [comment]
See #96 for another approach.

**Comment by jgm** [comment]
I've closed #71 after merging it into the `morecbits` branch in this repository.
It builds fine, we just need to change it so that the cbits are automatically generated.

**Comment by jgm** [comment]
I've added a build script for @cartazio's cbits and committed to master.
Result is a decent speedup in profiling build (from 14+ minutes to a bit under 9 minutes on the machine I'm using now, macos with clang).
The TeX reader is still the biggest bottleneck, though.
